triggers/tests/scenarios/Rule.daml

```diff
@@ -53,6 +55,86 @@ trigger = Trigger with
   registeredTemplates = RegisteredTemplates [registeredTemplate @T]
   heartbeat = None
 
+data RetryState = RetryState
+  { pendingRetries : Map.Map CommandId (Int, [Command])
+  , queuedRetry : Optional (CommandId, [Command])
+  }
+
+data Lens t a = Lens { getter : t -> a, setter : (a -> a) -> t -> t }
+
+infixr 9 @@
+(@@) : Lens h n -> Lens n a -> Lens h a
+hn @@ na = Lens (na.getter . hn.getter) (hn.setter . na.setter)
+
+_1 : Lens (a, b) a
+_1 = Lens fst (\f (a, b) -> (f a, b))
+
+_2 : Lens (a, b) b
+_2 = Lens snd (\f (a, b) -> (a, f b))
+
+pendingRetries' : Lens RetryState (Map.Map CommandId (Int, [Command]))
+pendingRetries' = Lens pendingRetries (\f s -> s { pendingRetries = f s.pendingRetries })
+
+queuedRetry' : Lens RetryState (Optional (CommandId, [Command]))
+queuedRetry' = Lens queuedRetry (\f s -> s { queuedRetry = f s.queuedRetry })
+
+infix 4 .=, %=
+(.=) : ActionState s m => Lens s a -> a -> m ()
+(.=) Lens { setter } = modify . setter . const
+(%=) : ActionState s m => Lens s a -> (a -> a) -> m ()
+(%=) Lens { setter } = modify . setter
+
+use : (ActionState s m, Functor m) => Lens s a -> m a
+use Lens { getter } = getter <$> get
+
+bracketRetry : Int -> Lens s RetryState -> Trigger s -> Trigger s
+bracketRetry retries rs trigger =
+  let updateState msg = do
+        case msg of
+          MCompletion (Completion { commandId, status }) -> case status of
+            Succeeded {} -> rs @@ pendingRetries' %= Map.delete commandId
+            Failed {} -> rs %= \s ->
+              optional s (enqueueRetry s commandId)
+                $ Map.lookup commandId s.pendingRetries
+          MTransaction {} -> pure ()
+        trigger.updateState msg
+
+      rule party = do
+        retry <- use (rs @@ queuedRetry')
+        retry `forA_` \(oldCmdId, cmds) -> do
+          -- the Failed would have removed the 2nd argument from the internal
+          -- state; I have no way to rederive it
+          newCmdId <- emitCommands cmds []
+          -- fixing the command ID now simplifies the update with
+          -- commands-in-flight later
+          rs @@ pendingRetries' %= \m ->
+            optional m (\v -> Map.insert newCmdId v $ Map.delete oldCmdId m)
+              $ Map.lookup oldCmdId m
+          rs @@ queuedRetry' .= None
+
+        -- NB: might be worth *not* performing the rule on msg if a retry was
+        -- enqueued.  Consider that runTrigger only runs the rule on a Failed
+        -- completion, not a succeeded one, but we have "filtered out" the
+        -- failure in a sense.  On the other hand, commands in flight *has*
+        -- changed; the rule might react to that in some way.
+        trigger.rule party
```

Comment:
I think I don’t quite understand how this is supposed to be used. Consider a rule that calls `emitCommands` if the ACS is in a given state. If my command fails, I will still keep resubmitting it every time since the ACS doesn’t change. It looks like with this PR we now have two retry mechanisms:

1. The implicit one you get by virtue of being state-based. This is still unlimited.
2. The explicit one you’ve added here. This one is limited.

What I had in mind for the issue was something that allows you to limit 1. Consider a buggy trigger which will try to submit a command which will reproducibly trigger (no pun intended) an assertion failure. With the current API you will basically DoS the ledger and send new commands as fast as you get failed completions and as a trigger author you don’t really have a way to avoid this (at least not when I wrote this issue). Some way of limiting retries of all or some commands could help here. Another option would be to not limit retries and instead go for #7044.

Does that make sense? Apologies if the issue was worded confusingly.

Reply:
It makes sense, and indeed is not at all what I get from #5978. 

Not invoking the rule on a failure, as mentioned in the comment, would be nearly good enough to do this. However, the rule might have saved command IDs in state, and we can't fix those up with the resubmissions.

I think what you describe is only possible to do at the user level with wrapper functions for emitting commands. I'll see how that can be done.